### PR TITLE
pytorch.org readme update

### PR DESCRIPTION
**Pull request by msaroufim**
Documentation change to pytorch.org including cherry picking adding more detail around CPU optimizations cfe6f101e47fa7a98efdfa0108666fe97b9614c0

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -224,7 +224,10 @@ High level performance data like Throughput or Percentile Precision can be gener
...
 *Note: the following property has bigger impact under heavy workloads.*
-If TorchServe is hosted on a machine with GPUs, there is a config property called `number_of_gpu` that tells the server to use a specific number of GPUs per model. In cases where we register multiple models with the server, this will apply to all the models registered. If this is set to a low value (ex: 0 or 1), it will result in under-utilization of GPUs. On the contrary, setting to a high value (>= max GPUs available on the system) results in as many workers getting spawned per model. Clearly, this will result in unnecessary contention for GPUs and can result in sub-optimal scheduling of threads to GPU.
+
+**CPU**: there is a config property called `workers` which sets the number of worker threads for a model. The best value to set `workers` to is 1 for CPU inference, otherwise the worker threads will contend with each other causing unncessary context switching. The best way to achieve paralellism on CPU is to then increase the `batch_size` parameter until diminishing returns are observed.
```
Comment (chauhang) [inline README.md:228]: Did we confirm this in our tests? Wasn't one worker per CPU core giving better results?

File: docs/README.md [hunk 2]
```diff
@@ -1,35 +1,44 @@
...
-## Advanced Features
-
-* [Advanced configuration](configuration.md) - Describes advanced TorchServe configurations.
-* [Custom Service](custom_service.md) - Describes how to develop custom inference services.
-* [Unit Tests](https://github.com/pytorch/serve/tree/master/ts/tests#testing-torchserve) - Housekeeping unit tests for TorchServe.
-* [Benchmark](https://github.com/pytorch/serve/tree/master/benchmarks#torchserve-model-server-benchmarking) - Use JMeter to run TorchServe through the paces and collect benchmark data.
-* [TorchServe on Kubernetes](https://github.com/pytorch/serve/blob/master/kubernetes/README.md#torchserve-on-kubernetes) -  Demonstrates a Torchserve deployment in Kubernetes using Helm Chart.
 
-## Default Handlers
+## Examples
```
Comment (chauhang) [inline docs/README.md:21]: Need to keep "Default Handlers" section for the built in handlers. There can be a separate section for Examples at the end

File: docs/README.md [hunk 3]
```diff
@@ -1,35 +1,44 @@
...
+* [HuggingFace Language Model](https://github.com/pytorch/serve/blob/master/examples/Huggingface_Transformers/Transformer_handler_generalized.py) - This handler takes an input sentence and can return sequence classifications, token classifications or Q&A answers
+* [Multi Modal Framework](https://github.com/pytorch/serve/blob/master/examples/MMF-activity-recognition/handler.py) - Build and deploy a classifier that combines text, audio and video input data
+* [Dual Translation Workflow](https://github.com/pytorch/serve/tree/master/examples/Workflows/nmt_tranformers_pipeline) - 
+* [Model Zoo](model_zoo.md) - List of pre-trained model archives ready to be served for inference with TorchServe.
+* [Examples](https://github.com/pytorch/serve/tree/master/examples) - Many examples of how to package and deploy models with TorchServe
+* [Workflow Examples](https://github.com/pytorch/serve/tree/master/examples/Workflows) - Examples of how to compose models in a workflow with TorchServe
+
+## Advanced Features
```
Comment (chauhang) [inline docs/README.md:34]: We also need separate page for Workflow and link here under Advanced features.

File: docs/README.md [hunk 4]
```diff
@@ -1,35 +1,44 @@
...
 
 ## Basic Features
 
 * [Serving Quick Start](https://github.com/pytorch/serve/blob/master/README.md#serve-a-model) - Basic server usage tutorial
 * [Model Archive Quick Start](https://github.com/pytorch/serve/tree/master/model-archiver#creating-a-model-archive) - Tutorial that shows you how to package a model archive file.
 * [Installation](https://github.com/pytorch/serve/blob/master/README.md#install-torchserve) - Installation procedures
-* [Serving Models](server.md) - Explains how to use torchserve
-   * [REST API](rest_api.md) - Specification on the API endpoint for TorchServe
+* [Serving Models](server.md) - Explains how to use TorchServe
+* [REST API](rest_api.md) - Specification on the API endpoint for TorchServe
```
Comment (chauhang) [inline docs/README.md:11]: Add links for the gRPC API as well

**Review by chauhang — COMMENTED** [review]
Please see comments inline

File: README.md [hunk 5]
```diff
@@ -224,7 +224,10 @@ High level performance data like Throughput or Percentile Precision can be gener
...
+
+**CPU**: there is a config property called `workers` which sets the number of worker threads for a model. The best value to set `workers` to is 1 for CPU inference, otherwise the worker threads will contend with each other causing unncessary context switching. The best way to achieve paralellism on CPU is to then increase the `batch_size` parameter until diminishing returns are observed.  
+
+**GPU**: there is a config property called `number_of_gpu` that tells the server to use a specific number of GPUs per model. In cases where we register multiple models with the server, this will apply to all the models registered. If this is set to a low value (ex: 0 or 1), it will result in under-utilization of GPUs. On the contrary, setting to a high value (>= max GPUs available on the system) results in as many workers getting spawned per model. Clearly, this will result in unnecessary contention for GPUs and can result in sub-optimal scheduling of threads to GPU.
```
Comment (HamidShojanazeri) [inline README.md:230]: @msaroufim I think in multi-gpu setting still we have control over to set the number of workers per model. Do we have any data on optimizing number of workers for GPUs.

Comment (msaroufim) [reply README.md:230]: I only have single GPU experiments where if you set num workers more than 1 it just crashes

Comment (msaroufim) [reply README.md:230]: On the num threads that is correct, it's why I abandoned the previous PR on setting torch num threads #1138

File: README.md (hunk 1 above)
Comment (msaroufim) [reply README.md:228]: That is not the case, row 25 to 46 makes this clear here https://drive.google.com/file/d/1m_I-ikRh_8i7s_ClLosyL-3whx6qCF5k/view?usp=sharing

File: docs/README.md (hunk 3 above)
Comment (msaroufim) [reply docs/README.md:34]: Ok we need to put a bit more thought into how we market it, I can take this on but I'd suggest we break into it its own PR

File: README.md (hunk 1 above)
Comment (HamidShojanazeri) [reply README.md:228]: @msaroufim if not misreading, it seems on rows 27-28 where its not optimized/ set_number_thread = default (if understood correctly) and workers=1, throughput is higher compared to 34-35 where set_number_thread=1 and workers=1 as well.

File: README.md (hunk 5 above)
Comment (HamidShojanazeri) [reply README.md:230]: > I only have single GPU experiments where if you set num workers more than 1 it just crashes

@msaroufim 
this happened in docker? we can sync off-line on this.

**Comment by msaroufim** [comment]
Reverted the CPU optimization cherry pick since it's best to discuss that seperately and very excited to to see this PR by @brianjo merged soon so we can have an actual table of contents https://github.com/brianjo/serve/commit/7c0ecc0b5688ee962423029b84331127675d3e4c

**Review by chauhang — APPROVED** [review]

**Review by HamidShojanazeri — APPROVED** [review]

**Pull request merged**